### PR TITLE
Add runtime-modes page

### PR DESCRIPTION
**Pull request by Zeegaan**
# Notes
- Implemented new Production modes page for version 10.1.0
- Added a link to runtime modes in index.md

### CODE REVIEW

**Comment by Zeegaan** [comment]
This will need an approving review from @ronaldbarendse before getting merged 🚀

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (sophie-neale) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:11]:
```suggestion
Runtime modes are settings you can enable via your Appsettings.json, which enforce best practices for different styles of development or production.
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (sophie-neale) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:21]:
```suggestion
BackofficeDevelopment mode is the default behavior for Umbraco and it does not enforce any settings at all.
```

**Review by sophie-neale — COMMENTED** [review]
Hey @Zeegaan! Thanks so much for adding this! When I meet with the Docs Curator team on Monday we'll go through it 👍🏼 but for now, I've added some small suggestions 🙂

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (ronaldbarendse) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:83]:
```suggestion

Although you can still edit document types and views (if not running from the published output), changes won't be picked up until you've rebuilt your project or republished the application. Also note that models won't be generated by ModelsBuilder (because the mode is set to `Nothing`), requiring your to do all your changes while in `Development` mode.

Also ensure the `UmbracoApplicationUrl` is updated to the primary URL of your production environment, as this is used when sending emails (password reset, notifications, health check results, etc.) and the keep-alive task.

## Customize/extend runtime mode validation

Validation of the above mentioned settings are done when determining the runtime level during startup using the new `IRuntimeModeValidationService` and when it fails, causes a `BootFailedException` to be thrown. The default implementation gets all registered `IRuntimeModeValidators` to do the validation, making it possible to remove default checks and/or add your own (inherit from `RuntimeModeProductionValidatorBase` if you only want to validate against the production runtime mode). The following validators are added by default:
- `JITOptimizerValidator` - Ensure the application is built/published in Release mode (with JIT optimization enabled) when in production runtime mode, e.g. using `dotnet publish --configuration Release`;
- `UmbracoApplicationUrlValidator` - ensure `Umbraco:CMS:WebRouting:UmbracoApplicationUrl` is configured when in production runtime mode;
- `UseHttpsValidator` - ensure `Umbraco:CMS:Global:UseHttps` is enabled when in production runtime mode;
- `RuntimeMinificationValidator` - ensure `Umbraco:CMS:RuntimeMinification:CacheBuster` is set to a fixed cache buster like `Version` or `AppDomain` when in production runtime mode;
- `ModelsBuilderModeValidator` - ensure `Umbraco:CMS:ModelsBuilder:ModelsMode` is not set to `InMemoryAuto` when in development runtime mode and set to `Nothing` when in production runtime mode.
```

The last part can be updated when PR https://github.com/umbraco/Umbraco-CMS/pull/12693 is merged, making it a much nicer experience to remove default runtime mode validators.

**Review by ronaldbarendse — CHANGES_REQUESTED** [review]
Thanks for the initial write-up @Zeegaan! I've rewritten large parts though to make the intent and usage of this new setting a bit more clear, please let me know if you think so as well 😇

I've also created PR https://github.com/umbraco/Umbraco-CMS/pull/12693 to make managing the validators a lot nicer: we should review and merge this before the final v10.1 is released if we want to avoid functional breaking changes (as it's only a different way of registration).

**Comment by Zeegaan** [comment]
@sophie-neale @eshanrnh this should now be ready for review! 💪

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:25]:
```suggestion
The `Development` mode can be used when you're developing from an IDE (like Visual Studio, VS Code, or Rider) or the dotnet CLI (e.g. using `dotnet watch`). It is a recommended prerequisite if you want to use the `Production` mode in your production environment.
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:21]:
```suggestion
The `BackofficeDevelopment` mode is the default behavior for Umbraco: it does not optimize Umbraco for any specific environment and does not have any prerequisites. This mode allows for rapid development (without having to recompile/rebuild your project), including all development from within the backoffice.
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:30]:
```suggestion
If you want to use the generated models, use `SourceCodeAuto` or `SourceCodeManual`, which requires manually recompiling the project after the models have changed (e.g. after updating Document Types, Media Types, Member Types, or Data Types). Razor views (`cshtml` files) will still be automatically compiled at runtime, allowing you to quickly iterate on the rendered output from templates, (macro) partial views, and view components.
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:49]:
```suggestion
Ensure your models are generated by running Umbraco and navigating to **Settings** > **Models Builder** > **Generate models**. You can remove the following properties from your `csproj` project file to enable the compilation of Razor views (which also ensures your views do not contain compilation errors and is a prerequisite for enabling `Production` mode):
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:57]:
```suggestion
Fix any compilation errors you might get after this, e.g. if you accidentally referenced deleted models or properties. Running the application will still show the rendered content and you're now ready to optionally enable `Production` mode on your production environment.
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:63]:
```suggestion
This mode disables both in-memory ModelsBuilder generation (see [Development mode](#development-mode)) and Razor (cshtml) runtime compilation. Production mode requires you to compile your views at build/publish time and enforces the following settings for optimal performance/security:
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:94]:
```suggestion
Although you can still edit document types and views (if not running from the published output), changes won't be picked up until you've rebuilt your project or republished the application. 

:::note
Models won't be generated by ModelsBuilder (because the mode is set to `Nothing`), requiring you to do all your changes while in `Development` mode.
:::
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (eshanrnh) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:100]:
```suggestion
Validation of the above-mentioned settings is done when determining the runtime level during startup using the new `IRuntimeModeValidationService` and when it fails, causes a `BootFailedException` to be thrown. The default implementation gets all registered `IRuntimeModeValidators` to do the validation, making it possible to remove default checks and/or add your own (inherit from `RuntimeModeProductionValidatorBase`, if you only want to validate against the production runtime mode). The following validators are added by default:
```

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (ronaldbarendse) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:54]: After some additional testing, the `CopyRazorGenerateFilesToPublishDirectory` property needs to be left enabled to ensure Deploy can generate signatures for templates (otherwise you'll get schema mismatches):
```suggestion
```

**Review by eshanrnh — COMMENTED** [review]
Looks good, @Zeegaan 🙌 Just a couple of suggestions around splitting long sentences for ease of readability and added a Note tag as per the [Markdown conventions](https://our.umbraco.com/Documentation/Contribute/Markdown-Conventions/#adding-notes-warnings-tips).

Let me know if something doesn't make sense 😀

File: Fundamentals/Setup/Server-Setup/runtime-modes.md
Comment (ronaldbarendse) [inline Fundamentals/Setup/Server-Setup/runtime-modes.md:109]: Now that PR https://github.com/umbraco/Umbraco-CMS/pull/12693 is merged, we can add the following sample code to show how you can remove default and/or add custom runtime mode validators:

```suggestion
- `ModelsBuilderModeValidator` - ensure `Umbraco:CMS:ModelsBuilder:ModelsMode` is not set to `InMemoryAuto` when in development runtime mode and set to `Nothing` when in production runtime mode.

The following example removes the default `UmbracoApplicationUrlValidator` and adds a new custom `DisableElectionForSingleServerValidator`:

    using System.Diagnostics.CodeAnalysis;
    using Microsoft.Extensions.Options;
    using Umbraco.Cms.Core.Composing;
    using Umbraco.Cms.Core.Configuration.Models;
    using Umbraco.Cms.Infrastructure.Runtime;
    using Umbraco.Cms.Infrastructure.Runtime.RuntimeModeValidators;
    
    public class RuntimeModeValidatorComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
            => builder.RuntimeModeValidators()
                .Remove<UmbracoApplicationUrlValidator>()
                .Add<DisableElectionForSingleServerValidator>();
    }
    
    public class DisableElectionForSingleServerValidator : IRuntimeModeValidator
    {
        private readonly IOptionsMonitor<GlobalSettings> _globalSettings;
    
        public DisableElectionForSingleServerValidator(IOptionsMonitor<GlobalSettings> globalSettings) => _globalSettings = globalSettings;
    
        public bool Validate(RuntimeMode runtimeMode, [NotNullWhen(false)] out string? validationErrorMessage)
        {
            if (_globalSettings.CurrentValue.DisableElectionForSingleServer == false)
            {
                validationErrorMessage = "Disable primary server election (and support for load balancing) to improve startup performance.";
                return false;
            }
    
            validationErrorMessage = null;
            return true;
        }
    }
```

Comment (ronaldbarendse) [reply Fundamentals/Setup/Server-Setup/runtime-modes.md:109]: I've indented the sample code 4 spaces (as mentioned in the Markdown conventions for our Docs), but other code blocks still use ` ```json` and ` ```xml` style markup...

Comment (eshanrnh) [reply Fundamentals/Setup/Server-Setup/runtime-modes.md:109]: @ronaldbarendse, you can either use 4 spaces or add an optional language identifier to enable syntax highlighting in your fenced code block. We might need to update the Markdown conventions article with this info 🙂

**Comment by eshanrnh** [comment]
Since @ronaldbarendse had requested some changes to this PR, can you approve it? The PR looks good to me.

Also, can this PR be merged immediately or should it be merged for a specific release (date)?

**Comment by ronaldbarendse** [comment]
@eshanrnh This looks great now and as it's already available in the 10.1.0-rc, so you can merge this right away!

**Pull request merged**
